File: 392_Is Subsequence.py
# 392. Is Subsequence

class Solution(object):
    def isSubsequence(self, s, t):
        """
        :type s: str
        :type t: str
        :rtype: bool
        """
        n = len(s)
        m = len(t)
        idx = 0
        d = dict()
        for i in range(n):
            if s[i] not in t[idx:]:
                return False
            if s[i] not in d:
                idx = t.index(s[i])
                d[s[i]] = idx
                idx += 1
            else:
                prev = d[s[i]]
                idx = max(prev+1,idx)
                if s[i] not in t[idx:]:
                    return False
                d[s[i]] = t[idx:].index(s[i])+prev+1

        return True

File: C++/392_is Subsequence.cpp
/*392.is Subsequence*/

class Solution {
public:
    bool isSubsequence(string s, string t) {
        int l1 = t.length();
        int l2 = s.length();
        int i = 0;
        int j = 0;
        while (i<l1){
            if (t[i] == s[j]){
                i += 1;
                j += 1;
            }
            else{
                i += 1;
            }

        }
        return j==l2;
    }
};
